File: lc/coreprep/implementation/patterns_reference_implementation/modified_binary_search/number_range.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

enum FindPos { MinPos, MaxPos };

int binary_search(const vector<int> &a, int low, int high, int k,
                  FindPos find) {
  int kpos = -1;

  while (low <= high) {
    int mid = low + (high - low) / 2;

    if (a[mid] < k) {
      low = mid + 1;
    } else if (a[mid] > k) {
      high = mid - 1;
    } else {
      kpos = mid;

      if (find == MinPos) {
        high = mid - 1;
      } else {
        low = mid + 1;
      }
    }
  }

  return kpos;
}

// O(logn) / O(1)
pair<int, int> find_min_max_pos(const vector<int> &a, int low, int high,
                                int k) {
  int min_pos = -1, max_pos = -1;

  min_pos = binary_search(a, low, high, k, MinPos);
  max_pos = binary_search(a, low, high, k, MaxPos);

  return make_pair(min_pos, max_pos);
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, k;
  cin >> tt;

  while (tt--) {
    cin >> n >> k;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    auto [min_pos, max_pos] = find_min_max_pos(a, 0, n - 1, k);
    cout << min_pos << " " << max_pos << "\n";
  }

  return 0;
}

File: lc/coreprep/implementation/patterns_reference_implementation/modified_binary_search/NumberRange.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class NumberRange {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        int k = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        int[] idx = findMinMax(a, k);
        System.out.printf("%d %d\n", idx[0], idx[1]);
      }
    }
  }

  static enum Find {
    MIN,
    MAX;
  }

  // O(logn) / O(1)
  public static int[] findMinMax(int[] a, int k) {
    int[] idx = new int[] {-1, -1};

    idx[0] = binarySearch(a, 0, a.length - 1, k, Find.MIN);
    idx[1] = binarySearch(a, 0, a.length - 1, k, Find.MAX);

    return idx;
  }

  private static int binarySearch(int[] a, int low, int high, int k,
                                  Find find) {
    int keyIndex = -1;

    while (low <= high) {
      int mid = low + (high - low) / 2;

      if (a[mid] < k) {
        low = mid + 1;
      } else if (a[mid] > k) {
        high = mid - 1;
      } else {
        keyIndex = mid;

        if (find == Find.MIN) {
          high = mid - 1;
        } else {
          low = mid + 1;
        }
      }
    }

    return keyIndex;
  }
}